evennia/commands/default/building.py

```diff
@@ -2868,51 +2896,45 @@ def func(self):
 
             results = ObjectDB.objects.filter(keyquery | aliasquery).distinct()
             nresults = results.count()
-
+            
+            # Use iterator to minimize memory ballooning on large result sets
+            results = results.iterator()
+            
             if nresults:
-                # convert result to typeclasses.
-                results = [result for result in results]
+                # filter results by typeclasses, if requested
+                obj_ids = []
                 if "room" in switches:
-                    results = [
-                        obj for obj in results if inherits_from(obj, ROOM_TYPECLASS)
-                    ]
+                    obj_ids.extend([
+                        obj.id for obj in results if inherits_from(obj, ROOM_TYPECLASS)
+                    ])
                 if "exit" in switches:
-                    results = [
-                        obj for obj in results if inherits_from(obj, EXIT_TYPECLASS)
-                    ]
+                    obj_ids.extend([
+                        obj.id for obj in results if inherits_from(obj, EXIT_TYPECLASS)
```

Comment:
There doesn't seem to be any unit tests for this, was this tested in practice? Because if both e.g. `room` and `exit` are in switches you'll be re-using the iterator (`results`) over and over, which is not safe. The generator is exhausted once you run it once. For example: 
```python
gen = evennia.ObjectDB.objects.all().iterator()

# using the generator for the first time
for o in gen: print(o)
Griatch
Limbo
box
...

# using the generator for the second time
for o in gen: print(o)
# no return since generator was exhausted already
```

The only way to handle this (as far as I know) is to re-create the generator every time you want to re-use it. Or turn it into a list, which is of course just what you wanted to avoid here.

Reply:
Heh, I had some doubts about what I did here as well-- for different reasons.

You are correct about the generator exhaustion (I was under the impression `.iterator()` was renewable), but even if it worked as I had anticipated it would still result in iterating the *entire* queryset up to 4 times to manually filter the results. Ouch.